### PR TITLE
unpin numpy and fix BasisState test

### PR DESCRIPTION
**Pull request by timmysilv**
this BasisState test fails because PL now uses the inputs as numpy array indices, so they should be integers. We can cast to ints on the PL side, but this should do for now. If people feel strongly about this casting, I can do it at the source (should be pretty straightforward)

### CODE REVIEW

**Review by mudit2812 — APPROVED** [review]
I don't have a strong opinion, but maybe wait for someone who might before merging

**Comment by timmysilv** [comment]
cc @trbromley

File: tests/test_gradients.py
Comment (trbromley) [inline tests/test_gradients.py:17]: Why does the unpinning lead to this issue? Or is this unrelated to the pinning?

Comment (trbromley) [reply tests/test_gradients.py:17]: Thanks! I have a preference to cast to ints internally if this is something we did before.

Comment (timmysilv) [reply tests/test_gradients.py:17]: [what we did before](https://github.com/PennyLaneAI/pennylane/blob/b7279db478fa144d6da91c1177e2845a21a28551/pennylane/devices/default_qubit_legacy.py#L793-L797) is completely different from [what we do now](https://github.com/PennyLaneAI/pennylane/blob/b7279db478fa144d6da91c1177e2845a21a28551/pennylane/ops/qubit/state_preparation.py#L118-L120), but floats would be accepted in the old way.

I'd still like to make this change here because `in_state` _should_ be integers. I'll cast to int in PennyLane itself as well, but I think this change makes sense regardless of the bug.

Comment (trbromley) [reply tests/test_gradients.py:17]: > I'll cast to int in PennyLane itself as well

Thanks @timmysilv!

Comment (trbromley) [reply tests/test_gradients.py:17]: This might already be causing an issue in the dev QML builds:
https://github.com/PennyLaneAI/qml/actions/runs/6306272465/job/17121149404#step:18:1593

This bumps up the priority of fixing the casting to help unblock the dev builds.

Comment (timmysilv) [reply tests/test_gradients.py:17]: full fix: PennyLaneAI/pennylane#4632

**Review by trbromley — APPROVED** [review]

**Pull request merged**
